### PR TITLE
Cut field underscore prefix

### PR DESCRIPTION
**Pull request by yaitskov**
# Description

Please include a summary of the change and which issue is fixed.

Fixes #(issue)

# Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added changes to the relevant changelog
- [x] I have made corresponding changes to the documentation
- [x] I have added haddock comments for every new function and data

### CODE REVIEW

**Comment by ludat** [comment]
@yaitskov can you rebase so the tests run? (the repo wasn't configured to run the tests before but now it should)

**Comment by yaitskov** [comment]
I pushed with force. Let's wait and see...

**Comment by yaitskov** [comment]
@ludat, seems ok.

File: example/config/development.json [hunk 1]
```diff
@@ -1 +1 @@
-{"warp": { "port": 9999}, "a": 7}
+{"server": { "port": 9999}, "a": 7}
```
Comment (ludat) [inline example/config/development.json:1]: I see now that the example is wrong, but I'd like this PR to only touch what it says it will touch, can you remove these changes? if you want to create a different PR that'd be nice, but I'll do it myself in the future anyhow.

File: example/src/Main.hs [hunk 2]
```diff
@@ -12,13 +12,13 @@ import GHC.Generics
...
   { appConfigServer :: Settings
-  , appConfigSeed :: Int
+  , _appConfig_seed :: Int
   } deriving (Generic)
 instance Conferer.FromConfig AppConfig
 instance Conferer.DefaultConfig AppConfig where
   configDef = AppConfig
     { appConfigServer = setPort 2222 Conferer.configDef -- If you want to configure new default for internal libs this is the place
-    , appConfigSeed = 17
+    , _appConfig_seed = 17
```
Comment (ludat) [inline example/src/Main.hs:21]: try to avoid changing files that are not directly related with what you want to change

File: packages/conferer/package.yaml [hunk 3]
```diff
@@ -5,6 +5,7 @@ synopsis: |
 description: |
   Library to abstract the parsing of many haskell config values from different config sources
 
+
```
Comment (ludat) [inline packages/conferer/package.yaml:8]: this whitespace changes nothing, can you remove it?

File: packages/conferer/src/Conferer/FromConfig/Internal.hs [hunk 4]
```diff
@@ -475,10 +478,13 @@ instance (IntoDefaultsG inner, Selector selector) =>
...
       fieldName = Text.pack $ selName @selector undefined
       prefix = applyFirst Char.toLower $ Text.pack s
+      underscorePrefix = "_" <> prefix <> "_"
       scopedKey =
         case Text.stripPrefix prefix fieldName of
           Just stripped -> applyFirst Char.toLower stripped
-          Nothing -> fieldName
+          Nothing -> case Text.stripPrefix underscorePrefix fieldName of
+            Just stripped -> stripped
+            Nothing -> maybe fieldName id $ Text.stripPrefix "_" fieldName
```
Comment (ludat) [inline packages/conferer/src/Conferer/FromConfig/Internal.hs:487]: nice :+1:

**Review by ludat — CHANGES_REQUESTED** [review]
A couple of things:

* Can you fill in the template? (description and completing the task list for the PR)
* Also can you add some tests to verify this behavior?
* Some tests are failing, I think that's related to the fact that we support old versions of ghc and with those `Data.Monoid.<>` wasn't part of prelude yet

File: packages/conferer/package.yaml (hunk 3 above)
Comment (yaitskov) [reply packages/conferer/package.yaml]: fixed

File: example/src/Main.hs (hunk 2 above)
Comment (yaitskov) [reply example/src/Main.hs]: I reverted, but I didn't get the issue. This is example project. Its goal to represent library features. I added new feature and showed how to use it. There are several field using original convention. I guess 1 field per notation is enough. Plus it helped me to test.

File: example/config/development.json (hunk 1 above)
Comment (yaitskov) [reply example/config/development.json]: reverted

**Review by ludat — CHANGES_REQUESTED** [review]
Very nice! the only thing missing are tests.

There is also the changelog and the docs but those are not documented anywhere so I can add them myself later

**Comment by ludat** [comment]
Sorry for taking so long but I just added a bunch of tests and a changelog so I'll merge this, thanks for contributing :clap:
